runner/shell.go

```diff
@@ -12,19 +13,44 @@ import (
 	"github.com/hashicorp/hcdiag/util"
 )
 
+type ShellConfig struct {
+	Command    string
+	Redactions []*redact.Redact
+	Timeout    time.Duration
+}
+
 // Shell runs shell commands in a real unix shell.
 type Shell struct {
+	ctx context.Context
+
 	Command    string           `json:"command"`
 	Shell      string           `json:"shell"`
 	Redactions []*redact.Redact `json:"redactions"`
+	Timeout    Timeout          `json:"timeout"`
 }
 
 // NewShell provides a runner for arbitrary shell code.
-func NewShell(command string, redactions []*redact.Redact) *Shell {
-	return &Shell{
-		Command:    command,
-		Redactions: redactions,
+func NewShell(cfg ShellConfig) (*Shell, error) {
+	return NewShellWithContext(context.Background(), cfg)
+}
+
+// NewShellWithContext provides a runner for arbitrary shell code.
+func NewShellWithContext(ctx context.Context, cfg ShellConfig) (*Shell, error) {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	timeout := cfg.Timeout
+	if timeout < 0 {
+		return nil, fmt.Errorf("timeout must be a nonnegative, timeout='%s'", timeout.String())
```

Comment:
I wouldn't block for this, but in Command and HTTP, I included config error types, which I thought could be useful for upstream type checking. Should we do the same for `Shell`? I would prefer if the foundational runners were consistent in this way, either with custom types or with a `fmt.Errorf`.

Reply:
I started implementing a `ShellConfigError` in this PR, but with just the timeout validation decided to skip the type. We definitely should go with a custom type in the future when there's more config validation here and/or we're specifically checking for Config error across multiple types of runners.